src/examples/AppUartModule.cpp

```diff
@@ -0,0 +1,270 @@
+                return;
+            }
+            if ((readBufferOffset == 0 && data->splitCount == 0) || (readBufferOffset > 0 && data->splitCount > 0)) {
+                CheckedMemcpy(&(readBuffer[readBufferOffset]), data->data, data->partLen);
+                readBufferOffset += data->partLen;
+
+                AppUartModuleTerminalResponseMessage message = { .commandSuccess = 1 };
+                SendModuleActionMessage(MessageType::MODULE_ACTION_RESPONSE, packet->header.sender,
+                    AppUartModuleActionResponseMessages::TERMINAL_RETURN_TYPE, 0, reinterpret_cast<u8*>(&message),
+                    SIZEOF_APP_UART_MODULE_TERMINAL_RESPONSE_MESSAGE, false);
+
+                if (data->splitHeader != MessageType::SPLIT_WRITE_CMD_END)
+                    return;
+#if IS_ACTIVE(LOGGING)
+                char temp[data->partLen + 1];
+                CheckedMemcpy(temp, data->data, data->partLen);
+                temp[data->partLen] = '\0';
+                logt("APPUART", "Got command one message with %s", temp);
+#endif
+                lineToReadAvailable = true;
+                GS->terminal.CheckAndProcessLine();
+                lineToReadAvailable = false;
+                readBufferOffset = 0;
+            }
+        }
+    }
+}
+
+NodeId AppUartModule::GetMeshAccessConnectionInNodeId(bool isVirtual) const
+{
+    MeshAccessConnections connIn = GS->cm.GetMeshAccessConnections(ConnectionDirection::DIRECTION_IN);
+    if (connIn.count < 1)
+        return NODE_ID_INVALID;
+    return isVirtual ? connIn.handles[0].GetVirtualPartnerId() : connIn.handles[0].GetPartnerId();
+}
+
+bool AppUartModule::IsConnectSmartPhone()
+{
+    const NodeId partnerId = GetMeshAccessConnectionInNodeId();
+    return partnerId >= NODE_ID_APP_BASE && partnerId < NODE_ID_GLOBAL_DEVICE_BASE;
+}
+
+void AppUartModule::SendAppLogQueue()
+{
+    if (logQueue._numElements < 2 || !IsConnectSmartPhone())
+        return;
+
+    AppUartLogRemain* logRemain = reinterpret_cast<AppUartLogRemain*>(logQueue.PeekNext().data);
+    char* log = reinterpret_cast<char*>(logQueue.PeekNext(1).data);
+    const u16 logLen = logRemain->logLen;
+    AppUartModuleTerminalLogMessage message;
+    if (logLen <= DATA_MAX_LEN) {
+        message.splitHeader = MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = 0;
+        message.partLen = logLen;
+        CheckedMemcpy(message.data, log, logLen);
+    }
+    const u16 remainLogLen = logLen - logRemain->sentLength;
+    if (logLen > DATA_MAX_LEN) {
+        message.splitHeader
+            = remainLogLen > DATA_MAX_LEN ? MessageType::SPLIT_WRITE_CMD : MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = (u8)(logRemain->sentLength / DATA_MAX_LEN);
+        message.partLen = remainLogLen > DATA_MAX_LEN ? DATA_MAX_LEN : remainLogLen;
+        CheckedMemcpy(message.data, &log[logRemain->sentLength], message.partLen);
+    }
+
+    const NodeId virtualPartnerId = GetMeshAccessConnectionInNodeId(true);
```

Comment:
Thanks for the great advice. I've added a feature to my smartphone application that sends a handshake message when a connection is established!

I'm just wondering how to handle critical sections and atomic variables, so I'd like your advice on my question above.

Reply:
Since your mentioning the app, when there are a lot of BLE-capable device around you (office environments e.g.) it is almost impossible to select a specific node in the app, since the reload will refresh the list every couple seconds. It would be perfect if there was a 'pause scanning' button which just halts the refreshing of the device list.

Otherwise very cool! :)

Reply:
And I almost forgot, a colleague who also tried the app mentioned that an option to sort by the connection strength / RSSI would really help identifying the device you want to connect to, at least in an example setting.

Reply:
Thanks for the feedback. I will add the ability to pause the scan and sort by RSSI strength.

Reply:
Scanning stops after 5 seconds, and tapping the refresh button (next to the filter button) on the application bar will scan again for 5 seconds.
Added the ability to filter by RSSI and to sort by RSSI. Please try [it](https://github.com/nishinohi/FruityMeshAppUart) out.